### ISSUE TITLE
Add assertion methods for floating point and integer format of `BigDecimal`

### ISSUE THREAD

**Issue by kiwi-oss** [assertj/assertj#3532]
#### Feature summary

A method `isInFloatingPointFormat` to assert that a `BigDecimal` has been specified in floating point syntax using a decimal dot or the exponential "e". 

A complementary method `isInIntegerFormat` asserting that the `BigDecimal` has been specified in an integer format should then also be added.

I first used `.scale().isNotZero()` as workaround, but the failure message is confusing because it refers to the scale, not the checked number. Adding `withFailMessage` isn't that useful as it doesn't allow printing the checked number.

Now I'm using `.is(new Condition<>(bd -> bd.scale() != 0, "in floating point format"))` and that works, but I think it makes sense to add a high-level assertion method that abstracts away the low-level check.

I have initially requested this for JsonUnit (lukas-krecan/JsonUnit#789), but it's more intuitive if the methods are added to AssertJ directly.

#### Example

```java
import static org.assertj.core.api.Assertions.assertThat;
import java.math.BigDecimal;
import java.util.stream.Stream;

var bigDecimals = Stream.of("653.401", "12.0", "10e3", "1.2e12").map(BigDecimal::new).toList();

// These should pass
assertThat(bigDecimals).allSatisfy(bd -> assertThat(bd).isInFloatingPointFormat());
assertThat(new BigDecimal("12")).isInIntegerFormat();

// These should fail
assertThat(bigDecimals).allSatisfy(bd -> assertThat(bd).isInIntegerFormat());
assertThat(new BigDecimal("12")).isInFloatingPointFormat();
```

**Comment by joel-costigliola** [comment]
That would be a good addition to avoid `new Condition<>(bd -> bd.scale() != 0`.

Are you keen to contribute it @kiwi-oss ?

**Comment by kiwi-oss** [comment]
Thanks for the feedback! My time is limited right now, so it'll take a while, but I'll give it a try

**Comment by joel-costigliola** [comment]
@kiwi-oss no rush, but let us know if you find after a while that you can't make it.

**Comment by JunHyungJang** [comment]
I also thought about the same issue and found this problem. Can I contribute to it?

**Comment by joel-costigliola** [comment]
@kiwi-oss let @JunHyungJang know if you can't contribute to this issue, he'll take over if you are ok with it.

**Comment by kiwi-oss** [comment]
Thanks for your offer, @JunHyungJang! Yes, you can take it over :slightly_smiling_face:

**Comment by joel-costigliola** [comment]
@kiwi-oss while I was reviewing @JunHyungJang PR, I have found the `isInIntegerFormat` assertion clear enough (that means it can be converted to an int) but `isInFloatingPointFormat` is a bit confusing, for example you say that `isInFloatingPointFormat(10e3)` should [pass](https://github.com/lukas-krecan/JsonUnit/issues/789#issuecomment-2213974425), 

When I see `10e3` I don't think floating point format but scientific format.

Having said that, I wonder if we should rename the assertion to something more obvious like `hasZeroScale` and `hasNonZeroScale`, there is no ambiguity there, it is crystal clear and you can get a decent error message showing the value and the scale (ex: "Expecting 10e3 to have a scale of zero but had a scale of -3".

Sorry @kiwi-oss and JunHyungJang, I should have looked into this more precisely !

What do you think guys ?

**Comment by kiwi-oss** [comment]
> I have found the isInIntegerFormat assertion clear enough (that means it can be converted to an int)

Maybe my suggested names weren't that clear after all because by adding the word "format" I meant to express that this is not about the type. It was supposed to designate any integer literal – it can be larger than fits into an `int`, `long` or other size-restricted type.

I agree that `10e3` being in a floating point format isn't very clear either.

Looking at these assertions from a JSON perspective, I wanted to avoid the term "scale" because I assume its meaning isn't clear to people who haven't read the Javadoc of `BigDecimal`. But these are `BigDecimal` assertions after all, so I think your suggestions make sense.

Maybe we could wrap the `BigDecimal` assertions in a "number" assertion class in JsonUnit and use less Java-specific wording there.

Thanks for having brought up the issue and sorry for the confusion!

**Comment by joel-costigliola** [comment]
After giving it some thoughts, I feel that `isInFloatingPointFormat` is too confusing to add it and `isInIntegerFormat` can be actually replaced by `hasScaleOf(0)`, so let's not add them for the time being until we find a better alternative (I'm still open to discussion and new suggestions).

On a note, you might have done already but going the `Condition` way is good, you can have a class with a static method factory to create the condition and use it with static import to have a nice readable assertion, something like:

```java
public static Condition<BigDecimal> InIntegerFormat = new Condition<>(bd -> bd.scale() == 0, "in integer point format");
```

```java
static import .... ExtraConditions.InIntegerFormat;
... 
assertThat(new BigDecimal("12")).is(InIntegerFormat());
```

It's quite close to `assertThat(new BigDecimal("12")).isInIntegerFormat();` just a pair of extra parentheses.

I'll leave the issue open for discussion and I'll close it after a while if we don't find better options.

**Comment by kiwi-oss** [comment]
Indeed, you can use `hasScaleOf(0)`. My problem was rather that I need the corresponding `doesNotHaveScaleOf(0)` and this doesn't exist.

So if we don't want to use your previous suggestion of `hasZeroScale` and `hasNonZeroScale`, I'd add at least `doesNotHaveScaleOf` – if only for consistency because most assertions have a negated counterpart if I'm not mistaken.

I do use a static method currently, but I found it incoherent to be forced to write it just because the positive assertion method exists, but not the negative one.

**Comment by joel-costigliola** [comment]
I can live with adding `doesNotHaveScaleOf`

**Comment by joel-costigliola** [comment]
An alternative way of solving would be to set some default description when navigating to `scale`, I just did a quick try and for this assertion: 

```java
assertThat(new BigDecimal(653)).scale().isNotZero();
```
I can get this error message:
```
java.lang.AssertionError: [asserting on scale of 653] 
Expecting actual:
  0
not to be equal to:
  0
```
Not as good as a tailored error message with the `doesNotHaveScaleOf` assertion but at least address your initial comment about the loss of context information.

**Comment by joel-costigliola** [comment]
The thing we want to avoid is to have to add every "integer" assertion to BigDecimal scale, that's too much work and is bloating the API, this is why navigation method like `scale()` are handy but we shoud make them better and set some navigation information on the chained assertions.

**Comment by kiwi-oss** [comment]
I agree that this would also be solved if the `scale()` navigation method provided the context, and it makes sense in general.

When this is implemented, we should remove the `hasScaleOf` method, in my opinion. This would remove both the redundancy with `scale().isEqualTo()` and the inconsistency of the missing `doesNotHaveScaleOf` method

**Comment by JunHyungJang** [comment]
I understand the point. It is quite ambiguous to use the isInFloatingPointFormat in 10e3 as you mentioned. Using scale looks more proper to use!
